src/patroni.py

```diff
@@ -61,9 +79,52 @@ def get_postgresql_state(self) -> str:
         Returns:
             running, restarting or stopping.
         """
-        r = requests.get(f"http://{self._pod_ip}:8008/health")
+        r = requests.get(f"http://{self._endpoint}:8008/health")
         return r.json()["state"]
 
+    @property
+    @retry(stop=stop_after_attempt(3), wait=wait_exponential(multiplier=1, min=2, max=10))
+    def cluster_members(self) -> set:
+        """Get the current cluster members."""
+        # Request info from cluster endpoint (which returns all members of the cluster).
+        r = requests.get(f"http://{self._endpoint}:8008/cluster")
+        return set([member["name"] for member in r.json()["members"]])
+
+    def are_all_members_ready(self) -> bool:
+        """Check if all members are correctly running Patroni and PostgreSQL.
+
+        Returns:
+            True if all members are ready False otherwise. Retries over a period of 10 seconds
+            3 times to allow server time to start up.
+        """
+        # Request info from cluster endpoint
+        # (which returns all members of the cluster and their states).
+        try:
+            for attempt in Retrying(stop=stop_after_delay(10), wait=wait_fixed(3)):
+                with attempt:
```

Comment:
is it better to use tenacity's retry decorator here?

Reply:
decorators are the things that go ontop of functions right?

If so, its better to have it nested in the code here since when it is ontop of the function it raises a `RetryError`  when it cannot complete the operation. This function is supposed to return false when the function fails, so this gaurentees it will return false instead of raising a `RetryError`

Reply:
I understood it the way Mia commented. Other approach would be to have two functions. This one with the retry decorator and another one that would do the same that was being done before (checking if a RetryError was thrown and set the return to False, which would be used in the charm code). This approach of only one function seems better to me.